rfcs/rfc0023-assignment-context.md

```diff
@@ -0,0 +1,179 @@
+# Assignemnt context attributes
+
+## Preamble
+
+    Author:  Curtis "Ovid" Poe <curtis.poe@gmail.com>
+    Sponsor:
+    ID:      0023
+    Status:  Draft
+
+## Abstract
+
+Add `:scalar`, `:list`, and `:void` attributes for subroutines/methods
+to ensure their results are handled correctly rather than relying on the
+`wantarray` heuristc.
+
+## Motivation
+
+* Create attributes for subroutines which require their return values to be
+  immediately assigned or ignored.
+* Not handling the context correctly will be a fatal error
+* Reduce dependency on fragile `wantarray` heuristics.
+* Reduce boilerplate code being hand-written over and over. 
+
+## Rationale
+
+It's very easy to write `wantarray` code that doesn't really do what you
+want. For example, I might have code which returns data which must be handled,
+so I use `wantarray` to guarantee that it's not called in void context:
+
+    sub munginator {
+        if ( !defined wantarray ) {
+            die "Do not call me in void context";
+        }
+        ...
+        return $some_scalar;
+    }
+
+The intent is that this should `die`:
+
+    my $value = get_value();
+    munginator($value); # fatal error!
+
+And this should work:
+
+    my $value = get_value();
+    $value = munginator($value);
+
+But here we have an edge case:
+
+    if ( munginator($value) ) {
+        ...
+    }
+
+That's evaluated in scalar context, so our `wantarray` heuristic has failed.
+We would like to not only stop reliance on a dodgy heuristic, but to not have
+to keep writing extra code to handle this case.
+
+Instead, it would be nice to write the following:
+
+    sub munginator :scalar {
+        ...
+        return $some_scalar;
+    }
+
+Any code calling `munginator` must do so by assigning the return value to a
+scalar. Otherwise the `if ( munginator($value) ) {...}` code fails as
+intended.
+
+## Specification
+
+Three new attributes are added, enabled via a feature named
+`assignment_context` (a better name would be good).
+
+1. `:scalar` - Return value must be assigned to a scalar, in scalar context
+2. `:list` - Return value must be assigned to a list
+3. `:void` - Return value must not be assigned to anything
```

Comment:
Oh. I see. That's entirely different to the idea I had in mind. What I was thinking was more that the attribute will force the context of a `return` expression or the final statement of the code body, into always being that context rather than determined at runtime based on the caller. In addition, the compiler can emit warnings (or errors), or the runtime can do so, if a function is called from a context that would be incompatible.

An "incompatible" context has to be a bit flexible though. It's perfectly fine to invoke a `:scalar` function from a list context - just remember it yields exactly one value. This is often what you want and is often fine - don't forget the list context of a function call's arguments:

```perl
sub get_thingy :scalar { ... }
SomeClass->new( thingy => get_thingy() );   ## this should be fine and allowable
```

All of these would then have nothing to do with the actual operation of assignments - that's different.

We could still as a separate idea (either in a new RFC or as part of this one) steal a function attribute from C compilers, which creates a warning if a function's result is unused; effectively, warn if invoked in void context. Sometimes it's OK - sometimes functions return an *optional* result value but it's semantically fine for the caller to ignore that, as it has done whatever work it needed to do as a side effect. At other times, it's very important that caller uses the value returned; calling the function in void context should be considered an error:

```perl
sub get_thingy :scalar :warn_unused { ... }

my $thing = get_thingy();  ## is fine
get_thingy(); 1;  ## warns (or fails to compile?)
```

Reply:
As with @leonerd, I'd want to preserve the sanctity of:

```perl
sub get_thingy :scalar { ... }
SomeClass->new( thingy => get_thingy() );
```

And also, this should work:

```perl
sub get_thingy :scalar { ... }
sub another_thingy :scalar { ... }
my $result = get_thingy() + another_thingy();
```

Reply:
OK, this seems reasonable. If I update the specification to state that `:scalar` succeeds even if called in list context, it seems like it would satisfy these concerns, but what about the propagation issue?

```perl
foo();
sub munginator :scalar {...}
sub foo {
    return munginator(42);
}
```

Since the above ultimately is in `:void` context, it still fails, but if assigned to a value at the very end, it succeeds, yes? This means that "assignment context" is a bad name for this feature.

As for @leonerd's question:

```perl
sub get_thingy :scalar :warn_unused { ... }

my $thing = get_thingy();  ## is fine
get_thingy(); 1;  ## warns (or fails to compile?)
```

I don't think we need `:warn_unused` unless we want this as a migration tool to have warnings. Then, developers can later remove the `:warn_unused` to get fatal behavior, but at their convenience. But then, we want this for the other contexts, yes?

```perl
sub get_thingy :scalar(warn) { ... }
```

Reply:
I like what @leonerd said, but wouldn't it be better to move it one step further?
`:scalar` would require an explicit return with a single scalar, `:list` with a list. `:void` would return nothing, but bare `return;` is allowed to exit early.

I feel like this is something worth having and would likely feel odd if it involved an extra attribute (so two in total).

Reply:
Not necessarily. A single scalar is a list in list context, and an array is a single value (its size) in scalar context. These are perfectly reasonable return values, as is an empty list (nothing) in list context.

Reply:
There are two possible useful features here: coercion of return context, and warning/error on strictly incompatible return context. Let's not introduce anything unperlish further than that.

Reply:
@Grinnz can you provide code examples of what you mean by this?

> There are two possible useful features here: coercion of return context.

Reply:
```
sub get_thingy :scalar { return; }
sub get_count :scalar { return @things; }

my %things = (thingy => get_thingy(), count => get_count());
```

effectively avoiding the injection vulnerability of using functions in an unexpected context.